File: apollo-runtime/src/main/java/com/apollographql/apollo/http/HttpExecutionContext.java
package com.apollographql.apollo.http;

import com.apollographql.apollo.api.ExecutionContext;
import com.apollographql.apollo.api.internal.Utils;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;

/**
 * Http GraphQL execution context, provides access to the raw {@link okhttp3.Response} response.
 */
public class HttpExecutionContext implements ExecutionContext.Element {

  public static final ExecutionContext.Key<HttpExecutionContext> KEY = new ExecutionContext.Key<HttpExecutionContext>() {
  };

  /**
   * Raw OkHttp response.
   */
  public final Response response;

  public HttpExecutionContext(@NotNull Response response) {
    this.response = Utils.checkNotNull(response, "response == null");
  }
}

File: apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo/network/HttpExecutionContext.kt
package com.apollographql.apollo.network

import com.apollographql.apollo.api.ApolloExperimental
import com.apollographql.apollo.api.ExecutionContext

@ApolloExperimental
sealed class HttpExecutionContext {

  data class Request(val headers: Map<String, String>) : ExecutionContext.Element {
    override val key: ExecutionContext.Key<*> = Key

    companion object Key : ExecutionContext.Key<Request>
  }

  data class Response(
      val statusCode: Int,
      val headers: Map<String, String>
  ) : ExecutionContext.Element {
    override val key: ExecutionContext.Key<*> = Key

    companion object Key : ExecutionContext.Key<Response>
  }
}
